sentryflow/pkg/receiver/svcmesh/istio/sidecar/sidecar.go

```diff
@@ -0,0 +1,536 @@
+	istionet "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/5GSEC/SentryFlow/pkg/config"
+	"github.com/5GSEC/SentryFlow/pkg/util"
+)
+
+const (
+	FilterName                  = "http-filter"
+	UpstreamAndClusterName      = "sentryflow"
+	ApiPath                     = "/api/v1/events"
+	RemoteWasmFilterClusterName = "remote_wasm"
+	Timeout                     = "5s"
+)
+
+// StartMonitoring begins monitoring API calls within the Istio (sidecar based)
+// service mesh deployed in a Kubernetes cluster. It achieves this by creating a
+// custom EnvoyFilter resource in Kubernetes.
+func StartMonitoring(ctx context.Context, cfg *config.Config, k8sClient client.Client) {
+	logger := util.LoggerFromCtx(ctx).Named("istio-sidecar")
+
+	logger.Info("Starting istio sidecar mesh monitoring")
+
+	if err := createEnvoyFilter(ctx, cfg, k8sClient); err != nil {
+		logger.Errorf("Failed to create EnvoyFilter. Stopping istio sidecar mesh monitoring, error: %v", err)
+		return
+	}
+	logger.Info("Started istio sidecar mesh monitoring")
+
+	<-ctx.Done()
+	logger.Info("Shutting down istio sidecar mesh monitoring")
+	if err := deleteEnvoyFilter(logger, k8sClient); err != nil {
+		logger.Errorf("Failed to delete EnvoyFilter, error: %v", err)
+	}
+
+	logger.Info("Stopped istio sidecar mesh monitoring")
+}
+
+func deleteEnvoyFilter(logger *zap.SugaredLogger, k8sClient client.Client) error {
+	existingFilter := &istionet.EnvoyFilter{}
+	if err := k8sClient.Get(context.Background(), types.NamespacedName{Name: FilterName, Namespace: "istio-system"}, existingFilter); err != nil {
+		return err
+	}
+
+	if err := k8sClient.Delete(context.Background(), existingFilter); err != nil {
+		return err
+	}
+	logger.Infow("Deleted EnvoyFilter", "name", FilterName, "namespace", "istio-system")
+
+	return nil
+}
+
+func createEnvoyFilter(ctx context.Context, cfg *config.Config, k8sClient client.Client) error {
+	logger := util.LoggerFromCtx(ctx)
+
+	var configVal = fmt.Sprintf(`{"upstream_name": "%v", "authority": "%v", "api_path": "%v"}
+`, UpstreamAndClusterName, UpstreamAndClusterName, ApiPath)
+
+	filter := &istionet.EnvoyFilter{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "EnvoyFilter",
+			APIVersion: "networking.istio.io/v1alpha3",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: FilterName,
+			// Deploy the filter to whatever istio considers its "root" namespace so that we
+			// don't have to create the ConfigMap(s) containing the WASM filter binary, and
+			// the associated annotations/configuration for the Istio sidecar(s).
+			// https://istio.io/latest/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig:~:text=No-,rootNamespace,-string
+			Namespace: "istio-system",
+			Labels: map[string]string{
+				"app.kubernetes.io/managed-by": "sentryflow",
+			},
+		},
+		Spec: networkingv1alpha3.EnvoyFilter{
```

Comment:
Do you think it's a better idea to marshal from a template string rather than using so much of nested structure, since it looks a bit unmaintainable this way. What do you think?

Reply:
You are correct. Maintaining this in the long run would be difficult. I have updated it to use a template string for better maintainability.